### DISCUSSION TITLE
Improve docs on behavour trees and builders

### DISCUSSION THREAD

**Discussion by maxsupermanhd** [adepierre/Botcraft#79] [Q&A] [answered]
I just can not understand what happened while I was not paying attention to the Botcraft.
What are those trees? Why trees? Why not just handlers of packets and simple loops?
What is this hell?! (<https://github.com/adepierre/Botcraft/blob/master/Examples/4_MapCreatorExample/src/main.cpp#L280>)
Where is good old onTick or onPacket?
What does selector, inverter, leaf, end and sequence do? Where can I find how to make my own trees?
In my opinion such trees add unnecessary complexity to the code, build up backtraces and require much more additional calls...

**Comment by adepierre** [comment, marked as answer]
Hi!

I think you've missed the wiki pages about this stuff. Their links are in the readme, but here is the one about [clients](https://github.com/adepierre/Botcraft/wiki/Clients) and [trees](https://github.com/adepierre/Botcraft/wiki/Behaviour-system). I'll answer your questions here too for completeness.

> what happened while I was not paying attention to the Botcraft

I kinda ... worked on it?


> What are those trees?

Behaviour trees are just mathematical models. They can be used to create complex behaviours from simple subtasks.


> Why trees?

Because of their shape. With node and leaf, they make think of trees.


> Why not just handlers of packets and simple loops?

Trees don't _replace_ packet handlers or loop or any other code. They are just a tool you can use or not on top of the rest of the library.


> What is this hell?! (https://github.com/adepierre/Botcraft/blob/master/Examples/4_MapCreatorExample/src/main.cpp#L280)

That's a full subtree that can be used when you need to eat, with many checks (set food in hand, try to eat, find food in nearby chest if no more in inventory etc...).


> Where is good old onTick or onPacket?

I'm not sure what you mean by onTick. Botcraft never had any onTick function. What you call onPacket is probably the ``Handle`` functions in Botcraft's clients, and these functions are still here.


> What does selector, inverter, leaf, end and sequence do? Where can I find how to make my own trees?

The wiki page mentioned above should answer that.


> In my opinion such trees add unnecessary complexity to the code, build up backtraces and require much more additional calls...

Your opinion is noted

**Comment by maxsupermanhd** [comment]
Oh I totally blinded that wiki pages that you made, perhaps will be good idea if you put those into readme or "home" maybe

**Reply by adepierre** [reply]
What do you mean in the readme? There are already mentioned with links to them in the readme. But it would be too long to copy all the content in the readme directly.
